grid_map_filters:

  # Fill holes in the map with inpainting.
  - name: inpaint
    type: gridMapCv/InpaintFilter
    params:
      input_layer: z
      output_layer: z_inpainted
      radius: 0.06

  # Reduce noise with a radial blurring filter.
  - name: mean_in_radius
    type: gridMapFilters/MeanInRadiusFilter
    params:
      input_layer: z_inpainted
      output_layer: z_smooth
      radius: 0.2

  # Compute surface normals.
  - name: surface_normals
    type: gridMapFilters/NormalVectorsFilter
    params:
      input_layer: z_smooth
      output_layers_prefix: normal_vectors_
      radius: 0.06
      normal_vector_positive_axis: z

  # Compute slope from surface normal.
  - name: slope
    type: gridMapFilters/MathExpressionFilter
    params:
      output_layer: slope
      expression: acos(normal_vectors_z)

  # Compute roughness as absolute difference from map to smoothened map.
  - name: roughness
    type: gridMapFilters/MathExpressionFilter
    params:
      output_layer: roughness
      expression: abs(z_inpainted - z_smooth)

  # Compute traversability as normalized weighted sum of slope and roughness.
  - name: traversability
    type: gridMapFilters/MathExpressionFilter
    params:
      output_layer: traversability
      expression: 0.5 * (1.0 - (slope / 0.6)) + 0.5 * (1.0 - (roughness / 0.1))

  # Set lower threshold on traversability.
  - name: traversability_lower_threshold
    type: gridMapFilters/ThresholdFilter
    params:
      condition_layer: traversability
      output_layer: traversability
      lower_threshold: 0.0
      set_to: 0.0

  # Set upper threshold on traversability.
  - name: traversability_upper_threshold
    type: gridMapFilters/ThresholdFilter
    params:
      condition_layer: traversability
      output_layer: traversability
      upper_threshold: 1.0
      set_to: 1.0 # Other uses: .nan, .inf